File: branches/AnyCpuNET4/Plugins/Kingbotk/AWB Plugin/WikiProjects/WPMilitaryHistory.vb
Friend NotInheritable Class WPMilitaryHistory
    Inherits PluginBase

    Private Const PluginName As String = "MilHist"
    Private Const TemplateName As String = "WPMILHIST"

    ' Initialisation:
    Friend Sub New()
        MyBase.New("WikiProject Military History") ' Specify alternate names only

        OurSettingsControl = New GenericWithWorkgroups(TemplateName, PluginName, False, True, params)
        OurSettingsControl.InspectUnsetText = "Remove importance"
    End Sub

    ' Settings:
    Private OurTab As New TabPage(PluginName)
    Private WithEvents OurSettingsControl As GenericWithWorkgroups

    Protected Friend Overrides ReadOnly Property PluginShortName() As String
        Get
            Return "MilitaryHistory"
        End Get
    End Property

    Protected Overrides ReadOnly Property PreferredTemplateName() As String
        Get
            Return TemplateName
        End Get
    End Property

    Protected Overrides Sub ImportanceParameter(ByVal Importance As Importance)
        ' WPMILHIST doesn't do importance
    End Sub

    Protected Friend Overrides ReadOnly Property GenericSettings() As IGenericSettings
        Get
            Return OurSettingsControl
        End Get
    End Property

    Protected Overrides ReadOnly Property CategoryTalkClassParm() As String
        Get
            Return "NA"
        End Get
    End Property

    Protected Overrides ReadOnly Property TemplateTalkClassParm() As String
        Get
            Return "NA"
        End Get
    End Property

    Friend Overrides ReadOnly Property HasReqPhotoParam() As Boolean
        Get
            Return False
        End Get
    End Property

    Friend Overrides Sub ReqPhoto()
    End Sub

    Const PeriodsAndConflictsGroup As String = "Periods and Conflicts"
    Const GeneralGroup As String = "General Task Forces"
    Const NationsGroup As String = "Nations and Regions"

    Dim params() As TemplateParameters =
    {
       New TemplateParameters() With {.StorageKey = "ACW", .Group = PeriodsAndConflictsGroup, .ParamName = "ACW"}, _
       New TemplateParameters() With {.StorageKey = "ARW", .Group = PeriodsAndConflictsGroup, .ParamName = "ARW"}, _
       New TemplateParameters() With {.StorageKey = "Classic", .Group = PeriodsAndConflictsGroup, .ParamName = "Classical"}, _
       New TemplateParameters() With {.StorageKey = "Crusades", .Group = PeriodsAndConflictsGroup, .ParamName = "Crusades"}, _
       New TemplateParameters() With {.StorageKey = "EarlyModern", .Group = PeriodsAndConflictsGroup, .ParamName = "Early-Modern"}, _
       New TemplateParameters() With {.StorageKey = "Medieval", .Group = PeriodsAndConflictsGroup, .ParamName = "Medieval"}, _
       New TemplateParameters() With {.StorageKey = "Muslim", .Group = PeriodsAndConflictsGroup, .ParamName = "Muslim"}, _
       New TemplateParameters() With {.StorageKey = "Napol", .Group = PeriodsAndConflictsGroup, .ParamName = "Napoleonic"}, _
       New TemplateParameters() With {.StorageKey = "WWI", .Group = PeriodsAndConflictsGroup, .ParamName = "WWI"}, _
       New TemplateParameters() With {.StorageKey = "WWII", .Group = PeriodsAndConflictsGroup, .ParamName = "WWII"}, _
       New TemplateParameters() With {.StorageKey = "Air", .Group = GeneralGroup, .ParamName = "Aviation"}, _
       New TemplateParameters() With {.StorageKey = "Biography", .Group = GeneralGroup, .ParamName = "Biography"}, _
       New TemplateParameters() With {.StorageKey = "Films", .Group = GeneralGroup, .ParamName = "Films"}, _
       New TemplateParameters() With {.StorageKey = "Fortifications", .Group = GeneralGroup, .ParamName = "Fortifications"}, _
       New TemplateParameters() With {.StorageKey = "Historiography", .Group = GeneralGroup, .ParamName = "Historiography"}, _
       New TemplateParameters() With {.StorageKey = "Intel", .Group = GeneralGroup, .ParamName = "Intel"}, _
       New TemplateParameters() With {.StorageKey = "LandVech", .Group = GeneralGroup, .ParamName = "Land Vehicles"}, _
       New TemplateParameters() With {.StorageKey = "Marit", .Group = GeneralGroup, .ParamName = "Martime"}, _
       New TemplateParameters() With {.StorageKey = "Memorial", .Group = GeneralGroup, .ParamName = "Memorials"}, _
       New TemplateParameters() With {.StorageKey = "National", .Group = GeneralGroup, .ParamName = "Nationals"}, _
       New TemplateParameters() With {.StorageKey = "Science", .Group = GeneralGroup, .ParamName = "Science"}, _
       New TemplateParameters() With {.StorageKey = "Tech", .Group = GeneralGroup, .ParamName = "Technology"}, _
       New TemplateParameters() With {.StorageKey = "Weapon", .Group = GeneralGroup, .ParamName = "Weaponry"}, _
       New TemplateParameters() With {.StorageKey = "NTF", .Group = NationsGroup, .ParamName = "No Task Force"}, _
       New TemplateParameters() With {.StorageKey = "African", .Group = NationsGroup, .ParamName = "Africa"}, _
       New TemplateParameters() With {.StorageKey = "Aus", .Group = NationsGroup, .ParamName = "Australia"}, _
       New TemplateParameters() With {.StorageKey = "Balkan", .Group = NationsGroup, .ParamName = "Balkan"}, _
       New TemplateParameters() With {.StorageKey = "Baltic", .Group = NationsGroup, .ParamName = "Baltic"}, _
       New TemplateParameters() With {.StorageKey = "Brit", .Group = NationsGroup, .ParamName = "British"}, _
       New TemplateParameters() With {.StorageKey = "Canuck", .Group = NationsGroup, .ParamName = "Canadian"}, _
       New TemplateParameters() With {.StorageKey = "China", .Group = NationsGroup, .ParamName = "Chinese"}, _
       New TemplateParameters() With {.StorageKey = "Dutch", .Group = NationsGroup, .ParamName = "Dutch"}, _
       New TemplateParameters() With {.StorageKey = "French", .Group = NationsGroup, .ParamName = "French"}, _
       New TemplateParameters() With {.StorageKey = "German", .Group = NationsGroup, .ParamName = "German"}, _
       New TemplateParameters() With {.StorageKey = "India", .Group = NationsGroup, .ParamName = "Indian"}, _
       New TemplateParameters() With {.StorageKey = "Italy", .Group = NationsGroup, .ParamName = "Italian"}, _
       New TemplateParameters() With {.StorageKey = "Japan", .Group = NationsGroup, .ParamName = "Japanese"}, _
       New TemplateParameters() With {.StorageKey = "Korean", .Group = NationsGroup, .ParamName = "Korean"}, _
       New TemplateParameters() With {.StorageKey = "Lebanese", .Group = NationsGroup, .ParamName = "Lebanese"}, _
       New TemplateParameters() With {.StorageKey = "MidEast", .Group = NationsGroup, .ParamName = "Middle-Eastern"}, _
       New TemplateParameters() With {.StorageKey = "NZ", .Group = NationsGroup, .ParamName = "New Zealand"}, _
       New TemplateParameters() With {.StorageKey = "Nordic", .Group = NationsGroup, .ParamName = "Nordic"}, _
       New TemplateParameters() With {.StorageKey = "Ottoman", .Group = NationsGroup, .ParamName = "Ottoman"}, _
       New TemplateParameters() With {.StorageKey = "Poland", .Group = NationsGroup, .ParamName = "Polish"}, _
       New TemplateParameters() With {.StorageKey = "Romanian", .Group = NationsGroup, .ParamName = "Romanian"}, _
       New TemplateParameters() With {.StorageKey = "Russian", .Group = NationsGroup, .ParamName = "Russian"}, _
       New TemplateParameters() With {.StorageKey = "Spanish", .Group = NationsGroup, .ParamName = "Spanish"}, _
       New TemplateParameters() With {.StorageKey = "SAmerican", .Group = NationsGroup, .ParamName = "S American"}, _
       New TemplateParameters() With {.StorageKey = "SEAsian", .Group = NationsGroup, .ParamName = "SE Asian"}, _
       New TemplateParameters() With {.StorageKey = "Taiwanese", .Group = NationsGroup, .ParamName = "Taiwanese"}, _
       New TemplateParameters() With {.StorageKey = "US", .Group = NationsGroup, .ParamName = "US"}
}

    Protected Friend Overrides Sub Initialise()
        OurMenuItem = New ToolStripMenuItem("Military History Plugin")
        MyBase.InitialiseBase() ' must set menu item object first
        OurTab.UseVisualStyleBackColor = True
        OurTab.Controls.Add(OurSettingsControl)
    End Sub

    ' Article processing:
    Protected Overrides ReadOnly Property InspectUnsetParameters() As Boolean
        Get
            Return OurSettingsControl.InspectUnsetParameters
        End Get
    End Property

    Protected Overrides Sub InspectUnsetParameter(ByVal Param As String)
        ' We only get called if InspectUnsetParameters is True
        If String.Equals(Param, "importance", StringComparison.CurrentCultureIgnoreCase) Then
            Article.DoneReplacement("importance=", "", True, PluginShortName)
        End If
    End Sub

    Protected Overrides Function SkipIfContains() As Boolean
    End Function

    Protected Overrides Sub ProcessArticleFinish()
        StubClass()
        With OurSettingsControl
            For Each lvi As ListViewItem In .ListView1.Items
                If lvi.Checked Then
                    Dim tp As TemplateParameters = DirectCast(lvi.Tag, TemplateParameters)
                    AddAndLogNewParamWithAYesValue(tp.ParamName.ToLower().Replace(" ", "-")) 'Probably needs some reformatting
                End If
            Next
        End With
        If Template.Parameters.ContainsKey("importance") Then
            Template.Parameters.Remove("importance")
            Article.ArticleHasAMajorChange()
            PluginManager.AWBForm.TraceManager.WriteArticleActionLine("Removed importance parameter", _
               PluginShortName)
        End If

        If Template.Parameters.ContainsKey("AncientNE") Then
            Template.Parameters.Remove("AncientNE")
            Article.ArticleHasAMajorChange()
            PluginManager.AWBForm.TraceManager.WriteArticleActionLine("Removed AncientNE parameter", _
               PluginShortName)
        End If

        If Template.Parameters.ContainsKey("auto") Then
            Template.Parameters.Remove("auto")
            Article.ArticleHasAMajorChange()
            PluginManager.AWBForm.TraceManager.WriteArticleActionLine("Removed auto parameter", _
               PluginShortName)
        End If
    End Sub

    Private Const conMedievalTaskForce As String = "Medieval-task-force"
    Protected Overrides Function TemplateFound() As Boolean
        Const conMiddleAges As String = "Middle-Ages-task-force"

        With Template
            If .Parameters.ContainsKey(conMiddleAges) Then
                If .Parameters(conMiddleAges).Value.ToLower = "yes" Then
                    .NewOrReplaceTemplateParm(conMedievalTaskForce, "yes", Article, False, False, False, "", _
                       PluginShortName)
                    Article.DoneReplacement(conMiddleAges, conMedievalTaskForce, True, PluginShortName)
                Else
                    Article.EditSummary += "deprecated Middle-Ages-task-force removed"
                    PluginManager.AWBForm.TraceManager.WriteArticleActionLine( _
                       "Middle-Ages-task-force parameter removed, not set to yes", PluginShortName)
                End If
                .Parameters.Remove(conMiddleAges)
                Article.ArticleHasAMinorChange()
            End If
        End With
    End Function

    Protected Overrides Sub GotTemplateNotPreferredName(ByVal TemplateName As String)
        ' Currently only WPBio does anything here (if {{musician}} add to musician-work-group)
    End Sub

    Protected Overrides Function WriteTemplateHeader(ByRef PutTemplateAtTop As Boolean) As String
        WriteTemplateHeader = "{{WPMILHIST" & _
           Microsoft.VisualBasic.vbCrLf & WriteOutParameterToHeader("class")
    End Function

    'User interface:
    Protected Overrides Sub ShowHideOurObjects(ByVal Visible As Boolean)
        PluginManager.ShowHidePluginTab(OurTab, Visible)
    End Sub

    ' XML settings:
    Protected Friend Overrides Sub ReadXML(ByVal Reader As System.Xml.XmlTextReader)
        Dim blnNewVal As Boolean = PluginManager.XMLReadBoolean(Reader, PluginName & "Enabled", Enabled)
        If Not blnNewVal = Enabled Then
            Enabled = blnNewVal ' Mustn't set if the same or we get extra tabs
        End If

        OurSettingsControl.ReadXML(Reader)
    End Sub

    Protected Friend Overrides Sub Reset()
        OurSettingsControl.Reset()
    End Sub

    Protected Friend Overrides Sub WriteXML(ByVal Writer As System.Xml.XmlTextWriter)
        Writer.WriteAttributeString(PluginName & "Enabled", Enabled.ToString)
        OurSettingsControl.WriteXML(Writer)
    End Sub
End Class

File: AWB/Plugins/Kingbotk Plugin/WikiProjects/WPMilitaryHistory.cs
using AutoWikiBrowser.Plugins.Kingbotk;
using AutoWikiBrowser.Plugins.Kingbotk.Plugins;
using System.Windows.Forms;
using System.Xml;

internal sealed class WPMilitaryHistory : PluginBase
{
    private const string PluginName = "MilHist";

    private const string TemplateName = "WPMILHIST";
    // Initialisation:
    internal WPMilitaryHistory() : base("WikiProject Military History")
    {
        // Specify alternate names only

        _ourSettingsControl = new GenericWithWorkgroups(TemplateName, PluginName, false, _params);
    }

    // Settings:
    private readonly TabPage _ourTab = new TabPage(PluginName);

    private readonly GenericWithWorkgroups _ourSettingsControl;

    protected internal override string PluginShortName
    {
        get { return "Military History"; }
    }

    protected override string PreferredTemplateName
    {
        get { return TemplateName; }
    }

    protected internal override IGenericSettings GenericSettings
    {
        get { return _ourSettingsControl; }
    }

    internal override bool HasReqPhotoParam
    {
        get { return false; }
    }

    internal override void ReqPhoto()
    {
    }

    private const string PeriodsAndConflictsGroup = "Periods and Conflicts";
    private const string GeneralGroup = "General Task Forces";

    private const string NationsGroup = "Nations and Regions";

    private readonly TemplateParameters[] _params =
    {
        new TemplateParameters
        {
            StorageKey = "ACW",
            Group = PeriodsAndConflictsGroup,
            ParamName = "ACW"
        },
        new TemplateParameters
        {
            StorageKey = "ARW",
            Group = PeriodsAndConflictsGroup,
            ParamName = "ARW"
        },
        new TemplateParameters
        {
            StorageKey = "Classic",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Classical"
        },
        new TemplateParameters
        {
            StorageKey = "Crusades",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Crusades"
        },
        new TemplateParameters
        {
            StorageKey = "EarlyModern",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Early-Modern"
        },
        new TemplateParameters
        {
            StorageKey = "Medieval",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Medieval"
        },
        new TemplateParameters
        {
            StorageKey = "Muslim",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Muslim"
        },
        new TemplateParameters
        {
            StorageKey = "Napol",
            Group = PeriodsAndConflictsGroup,
            ParamName = "Napoleonic"
        },
        new TemplateParameters
        {
            StorageKey = "WWI",
            Group = PeriodsAndConflictsGroup,
            ParamName = "WWI"
        },
        new TemplateParameters
        {
            StorageKey = "WWII",
            Group = PeriodsAndConflictsGroup,
            ParamName = "WWII"
        },
        new TemplateParameters
        {
            StorageKey = "Air",
            Group = GeneralGroup,
            ParamName = "Aviation"
        },
        new TemplateParameters
        {
            StorageKey = "Biography",
            Group = GeneralGroup,
            ParamName = "Biography"
        },
        new TemplateParameters
        {
            StorageKey = "Films",
            Group = GeneralGroup,
            ParamName = "Films"
        },
        new TemplateParameters
        {
            StorageKey = "Fortifications",
            Group = GeneralGroup,
            ParamName = "Fortifications"
        },
        new TemplateParameters
        {
            StorageKey = "Historiography",
            Group = GeneralGroup,
            ParamName = "Historiography"
        },
        new TemplateParameters
        {
            StorageKey = "Intel",
            Group = GeneralGroup,
            ParamName = "Intel"
        },
        new TemplateParameters
        {
            StorageKey = "LandVech",
            Group = GeneralGroup,
            ParamName = "Land Vehicles"
        },
        new TemplateParameters
        {
            StorageKey = "Marit",
            Group = GeneralGroup,
            ParamName = "Martime"
        },
        new TemplateParameters
        {
            StorageKey = "Memorial",
            Group = GeneralGroup,
            ParamName = "Memorials"
        },
        new TemplateParameters
        {
            StorageKey = "National",
            Group = GeneralGroup,
            ParamName = "Nationals"
        },
        new TemplateParameters
        {
            StorageKey = "Science",
            Group = GeneralGroup,
            ParamName = "Science"
        },
        new TemplateParameters
        {
            StorageKey = "Tech",
            Group = GeneralGroup,
            ParamName = "Technology"
        },
        new TemplateParameters
        {
            StorageKey = "Weapon",
            Group = GeneralGroup,
            ParamName = "Weaponry"
        },
        new TemplateParameters
        {
            StorageKey = "NTF",
            Group = NationsGroup,
            ParamName = "No Task Force"
        },
        new TemplateParameters
        {
            StorageKey = "African",
            Group = NationsGroup,
            ParamName = "Africa"
        },
        new TemplateParameters
        {
            StorageKey = "Aus",
            Group = NationsGroup,
            ParamName = "Australia"
        },
        new TemplateParameters
        {
            StorageKey = "Balkan",
            Group = NationsGroup,
            ParamName = "Balkan"
        },
        new TemplateParameters
        {
            StorageKey = "Baltic",
            Group = NationsGroup,
            ParamName = "Baltic"
        },
        new TemplateParameters
        {
            StorageKey = "Brit",
            Group = NationsGroup,
            ParamName = "British"
        },
        new TemplateParameters
        {
            StorageKey = "Canuck",
            Group = NationsGroup,
            ParamName = "Canadian"
        },
        new TemplateParameters
        {
            StorageKey = "China",
            Group = NationsGroup,
            ParamName = "Chinese"
        },
        new TemplateParameters
        {
            StorageKey = "Dutch",
            Group = NationsGroup,
            ParamName = "Dutch"
        },
        new TemplateParameters
        {
            StorageKey = "French",
            Group = NationsGroup,
            ParamName = "French"
        },
        new TemplateParameters
        {
            StorageKey = "German",
            Group = NationsGroup,
            ParamName = "German"
        },
        new TemplateParameters
        {
            StorageKey = "India",
            Group = NationsGroup,
            ParamName = "Indian"
        },
        new TemplateParameters
        {
            StorageKey = "Italy",
            Group = NationsGroup,
            ParamName = "Italian"
        },
        new TemplateParameters
        {
            StorageKey = "Japan",
            Group = NationsGroup,
            ParamName = "Japanese"
        },
        new TemplateParameters
        {
            StorageKey = "Korean",
            Group = NationsGroup,
            ParamName = "Korean"
        },
        new TemplateParameters
        {
            StorageKey = "Lebanese",
            Group = NationsGroup,
            ParamName = "Lebanese"
        },
        new TemplateParameters
        {
            StorageKey = "MidEast",
            Group = NationsGroup,
            ParamName = "Middle-Eastern"
        },
        new TemplateParameters
        {
            StorageKey = "NZ",
            Group = NationsGroup,
            ParamName = "New Zealand"
        },
        new TemplateParameters
        {
            StorageKey = "Nordic",
            Group = NationsGroup,
            ParamName = "Nordic"
        },
        new TemplateParameters
        {
            StorageKey = "Ottoman",
            Group = NationsGroup,
            ParamName = "Ottoman"
        },
        new TemplateParameters
        {
            StorageKey = "Poland",
            Group = NationsGroup,
            ParamName = "Polish"
        },
        new TemplateParameters
        {
            StorageKey = "Romanian",
            Group = NationsGroup,
            ParamName = "Romanian"
        },
        new TemplateParameters
        {
            StorageKey = "Russian",
            Group = NationsGroup,
            ParamName = "Russian"
        },
        new TemplateParameters
        {
            StorageKey = "Spanish",
            Group = NationsGroup,
            ParamName = "Spanish"
        },
        new TemplateParameters
        {
            StorageKey = "SAmerican",
            Group = NationsGroup,
            ParamName = "S American"
        },
        new TemplateParameters
        {
            StorageKey = "SEAsian",
            Group = NationsGroup,
            ParamName = "SE Asian"
        },
        new TemplateParameters
        {
            StorageKey = "Taiwanese",
            Group = NationsGroup,
            ParamName = "Taiwanese"
        },
        new TemplateParameters
        {
            StorageKey = "US",
            Group = NationsGroup,
            ParamName = "US"
        }
    };

    protected internal override void Initialise()
    {
        OurMenuItem = new ToolStripMenuItem("Military History Plugin");
        InitialiseBase();
        // must set menu item object first
        _ourTab.UseVisualStyleBackColor = true;
        _ourTab.Controls.Add(_ourSettingsControl);
    }

    // Article processing:
    protected override bool SkipIfContains()
    {
        return false;
    }

    protected override void ProcessArticleFinish()
    {
        StubClass();
        foreach (ListViewItem lvi in _ourSettingsControl.ListView1.Items)
        {
            if (lvi.Checked)
            {
                TemplateParameters tp = (TemplateParameters) lvi.Tag;
                AddAndLogNewParamWithAYesValue(tp.ParamName.ToLower().Replace(" ", "-"));
                //Probably needs some reformatting
            }
        }
        if (Template.Parameters.ContainsKey("importance"))
        {
            Template.Parameters.Remove("importance");
            TheArticle.ArticleHasAMajorChange();
        }

        if (Template.Parameters.ContainsKey("auto"))
        {
            Template.Parameters.Remove("auto");
            TheArticle.ArticleHasAMajorChange();
        }
    }

    private const string MedievalTaskForce = "Medieval-task-force";

    protected override bool TemplateFound()
    {
        const string conMiddleAges = "Middle-Ages-task-force";

        if (Template.Parameters.ContainsKey(conMiddleAges))
        {
            if (Template.Parameters[conMiddleAges].Value.ToLower() == "yes")
            {
                Template.NewOrReplaceTemplateParm(MedievalTaskForce, "yes", TheArticle, false, false, false, "",
                    PluginShortName);
                TheArticle.DoneReplacement(conMiddleAges, MedievalTaskForce);
            }
            else
            {
                TheArticle.EditSummary += "deprecated Middle-Ages-task-force removed";
            }
            Template.Parameters.Remove(conMiddleAges);
            TheArticle.ArticleHasAMinorChange();
        }
        return false;
    }

    protected override string WriteTemplateHeader()
    {
        return "{{WPMILHIST" + WriteOutParameterToHeader("class");
    }

    //User interface:
    protected override void ShowHideOurObjects(bool visible)
    {
        PluginManager.ShowHidePluginTab(_ourTab, visible);
    }

    // XML settings:
    protected internal override void ReadXML(XmlTextReader reader)
    {
        bool blnNewVal = PluginManager.XMLReadBoolean(reader, PluginName + "Enabled", Enabled);
        // ReSharper disable once RedundantCheckBeforeAssignment
        if (blnNewVal != Enabled)
        {
            Enabled = blnNewVal;
            // Mustn't set if the same or we get extra tabs
        }

        _ourSettingsControl.ReadXML(reader);
    }

    protected internal override void Reset()
    {
        _ourSettingsControl.Reset();
    }

    protected internal override void WriteXML(XmlTextWriter writer)
    {
        writer.WriteAttributeString(PluginName + "Enabled", Enabled.ToString());
        _ourSettingsControl.WriteXML(writer);
    }
}